daemon/logger/loggerutils/logfile.go

```diff
@@ -587,27 +785,48 @@ func tailFiles(files []SizeReaderAt, watcher *logger.LogWatcher, dec Decoder, ge
 		}
 	}()
 
-	readers := make([]io.Reader, 0, len(files))
+	readers, err := getTailFiles(ctx, files, nLines, getTailReader)
+	if err != nil {
+		watcher.Err <- err
+		return false
+	}
 
-	if nLines > 0 {
-		for i := len(files) - 1; i >= 0 && nLines > 0; i-- {
-			tail, n, err := getTailReader(ctx, files[i], nLines)
-			if err != nil {
-				watcher.Err <- errors.Wrap(err, "error finding file position to start log tailing")
-				return false
+	var idx int
+	defer func() {
+		// Make sure all are released if there is an early return.
+		if !cont {
+			for _, r := range readers[idx:] {
+				if err := r.Close(); err != nil {
+					log.G(ctx).WithError(err).Warn("Error closing log reader")
+				}
 			}
-			nLines -= n
-			readers = append([]io.Reader{tail}, readers...)
 		}
-	} else {
-		for _, r := range files {
-			readers = append(readers, r)
+	}()
+
+	for _, ra := range readers {
+		dec.Reset(ra)
+
+		err := fwd.Do(ctx, watcher, dec)
+		if errClose := ra.Close(); errClose != nil {
+			log.G(ctx).WithError(errClose).Warn("Error closing log reader")
+		}
+		idx++
+
+		if err != nil {
+			if errors.Is(err, errConsumerGone) {
+				span.AddEvent("consumer gone")
+				return false
+			}
+
+			if errors.Is(err, errLogTooNew) {
+				span.AddEvent("Remaining logs are newer than requested time range")
+				return false
+			}
+			log.G(ctx).WithError(err).Warn("Skipping remainder of log file due to error")
 		}
```

Comment:
This works for me.

I _think_ the reason I originally did this was because the decoder was returning an error with the bad offset, then we'd advance that offset in logfile.
This is of course not part of it anymore and as you pointed out we can probably deal with those type of errors directly in the decoder anyway.